File: app/src/main/java/se/gustavkarlsson/skylight/android/gui/activities/main/fragments/aurora_chance/AuroraChanceFragment.java
package se.gustavkarlsson.skylight.android.gui.activities.main.fragments.aurora_chance;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import javax.inject.Inject;
import javax.inject.Named;

import se.gustavkarlsson.skylight.android.R;
import se.gustavkarlsson.skylight.android.dagger.modules.replaceable.FragmentRootViewModule;
import se.gustavkarlsson.skylight.android.evaluation.Chance;
import se.gustavkarlsson.skylight.android.evaluation.ChanceEvaluator;
import se.gustavkarlsson.skylight.android.evaluation.ChanceLevel;
import se.gustavkarlsson.skylight.android.gui.AuroraReportUpdateListener;
import se.gustavkarlsson.skylight.android.gui.activities.main.MainActivity;
import se.gustavkarlsson.skylight.android.models.AuroraReport;

import static se.gustavkarlsson.skylight.android.dagger.modules.replaceable.FragmentRootViewModule.FRAGMENT_ROOT_NAME;

public class AuroraChanceFragment extends Fragment implements AuroraReportUpdateListener {
	private static final String TAG = AuroraChanceFragment.class.getSimpleName();

	@Inject
	ChanceEvaluator<AuroraReport> evaluator;

	@Inject
	@Named(FRAGMENT_ROOT_NAME)
	View rootView;

	@Inject
	LocationPresenter locationPresenter;

	@Inject
	TimeSinceUpdatePresenter timeSinceUpdatePresenter;

	@Inject
	ChancePresenter chancePresenter;

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		Log.v(TAG, "onCreateView");
		((MainActivity) getActivity()).getComponent()
				.getAuroraChanceFragmentComponent(new FragmentRootViewModule(inflater, container, R.layout.fragment_aurora_chance))
				.inject(this);
		return rootView;
	}

	@Override
	public void onStart() {
		Log.v(TAG, "onStart");
		super.onStart();
		timeSinceUpdatePresenter.onStart();
	}

	@Override
	public void onUpdate(AuroraReport report) {
		Log.v(TAG, "onUpdate");
		locationPresenter.onUpdate(report.getAddress());
		timeSinceUpdatePresenter.onUpdate(report.getTimestampMillis());
		Chance chance = evaluator.evaluate(report);
		chancePresenter.onUpdate(ChanceLevel.fromChance(chance));
		rootView.invalidate();
	}

	@Override
	public void onDestroyView() {
		Log.v(TAG, "onDestroyView");
		timeSinceUpdatePresenter.destroy();
		evaluator = null;
		rootView = null;
		locationPresenter = null;
		timeSinceUpdatePresenter = null;
		chancePresenter = null;
		super.onDestroyView();
	}
}

File: app/src/main/java/se/gustavkarlsson/skylight/android/gui/activities/main/fragments/aurora_chance/AuroraChanceFragment.kt
package se.gustavkarlsson.skylight.android.gui.activities.main.fragments.aurora_chance

import android.os.Bundle
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import org.threeten.bp.Instant
import se.gustavkarlsson.skylight.android.R
import se.gustavkarlsson.skylight.android.dagger.Names.FRAGMENT_ROOT_NAME
import se.gustavkarlsson.skylight.android.dagger.Names.LATEST_NAME
import se.gustavkarlsson.skylight.android.dagger.modules.replaceable.FragmentRootViewModule
import se.gustavkarlsson.skylight.android.gui.activities.main.MainActivity
import se.gustavkarlsson.skylight.android.models.AuroraReport
import se.gustavkarlsson.skylight.android.observers.ObservableValue
import se.gustavkarlsson.skylight.android.observers.ValueObserver
import javax.inject.Inject
import javax.inject.Named

class AuroraChanceFragment : Fragment(), ValueObserver<AuroraReport> {

    @Inject
    @field:Named(FRAGMENT_ROOT_NAME)
    lateinit var rootView: View

    @Inject
	lateinit var locationPresenter: LocationPresenter

    @Inject
	lateinit var timeSinceUpdatePresenter: TimeSinceUpdatePresenter

    @Inject
	lateinit var chancePresenter: ChancePresenter

    @Inject
    @field:Named(LATEST_NAME)
	lateinit var latestAuroraReport: ObservableValue<AuroraReport>

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        Log.v(TAG, "onCreateView")
        (activity as MainActivity).component
                .getAuroraChanceFragmentComponent(FragmentRootViewModule(inflater, container, R.layout.fragment_aurora_chance))
                .inject(this)
        return rootView
    }

    override fun onStart() {
        Log.v(TAG, "onStart")
        super.onStart()
        latestAuroraReport.addListener(this)
        updatePresenters(latestAuroraReport.value)
        timeSinceUpdatePresenter.start()
    }

    override fun valueChanged(newData: AuroraReport) {
        Log.v(TAG, "valueChanged")
        activity.runOnUiThread { updatePresenters(newData) }
    }

    private fun updatePresenters(report: AuroraReport) {
        locationPresenter.update(report.address)
        timeSinceUpdatePresenter.update(Instant.ofEpochMilli(report.timestampMillis))
        chancePresenter.update(report)
    }

    override fun onStop() {
        Log.v(TAG, "onStop")
        latestAuroraReport.removeListener(this)
        timeSinceUpdatePresenter.stop()
        super.onStop()
    }

    companion object {
        private val TAG = AuroraChanceFragment::class.java.simpleName
    }
}
